plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts

```diff
@@ -0,0 +1,112 @@
+    const { filepath } = parseGitUri(location.target);
+
+    // TODO: Maybe find a better way to detect whether discovery is necessary
+    //  (e.g. by trying to download it, or by reading the tree first?)
+    if (pattern.test(filepath ?? '')) {
+      return false;
+    }
+
+    try {
+      // TODO: This approach is using readTree to access the file (names) in the repository.
+      //  This can be quite slow in case the repository is very big. Especially as we don't need
+      //  the contents but only the file names! Alternatives would be
+      //  tree API from github (https://docs.github.com/en/free-pro-team@latest/rest/reference/git#get-a-tree)
+      //  however, even the recrusive mode there has limits. This could be a separate API.
+      //  This additional (for example listTree()) API could also be used in the catalog-import
+      //  plugin to check whether the repository already contains a catalog-info.yaml before
+      //  creating a PR for it.
+
+      // TODO: readTree doesn't work for a missing ref, so target has to look like
+      //  https://github.com/backstage/backstage/tree/master for now. On the other side
+      //  this is good, as it allows to work around the catalog-import plugin.
+      const tree = await this.options.reader.readTree(location.target, {
+        filter(path: string): boolean {
+          return pattern.test(path);
+        },
+      });
+      const files = await tree.files();
+
+      for (const file of files) {
+        emit({
+          type: 'location',
+          location: {
+            type: 'url',
+            target: await this.generateUrl(location.target, file.path),
+          },
+          optional: false,
+        });
+      }
+    } catch (error) {
+      const message = `Unable to read ${location.type}, ${error}`;
+
+      emit(result.generalError(location, message));
+    }
+
+    return true;
+  }
+
+  // TODO: This is provider specific! We need a way to generate the full
+  //  path to a file in a repository. It would be appealing if the return
+  //  value of readTree (or more specific of files() so part of ReadTreeResponseFile)
+  //  contains a url property with this url. A UrlReader should have all
+  //  details to generate it. However this feels like a breaking change
+  //  that all existing readers have to fill.
+  private async generateUrl(target: string, file: string): Promise<string> {
+    const gitUrl = parseGitUri(target);
+    const baseUrl = stringify(gitUrl);
+    // TODO: The GitHubReader doesn't know the ref yet if it's missing.
+    //  We could extend it by using the github API to retrieve the default
+    //  branch. On the other hand I'm not sure where the "..archive/${ref}.tar.gz"
+    //  url is comming from. The official endpoint says that REF is optional:
+    //  https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#download-a-repository-archive-tar
+    const ref = 'master';
+
+    return `${baseUrl}/blob/${ref}/${file}`;
```

Comment:
Elsewhere in the code base I think there are other references to it, but I believe this format only works on GitHub and possibly GitLab, but does NOT work on BitBucket?  Wonder if we need a utility function somehow to align finding a file in a branch by Git provider?  (even if a future iteration)

Reply:
Yeah, this is provide specific, see the TODO comments